### PR TITLE
BGL version for Linear_cell_complex_for_combinatorial_map

### PR DESCRIPTION
**Pull request by gdamiand**
_Please use the following template to help us managing pull requests._

## Summary of Changes

Add BGL support for Linear_cell_complex_for_combinatorial_map.
(cf small feature [BGL_for_lcc](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/BGL_for_lcc))

Note that several examples and tests are updated (the list is given in the small feature page).

## Release Management

* Affected package(s): Linear_cell_complex
* Feature/Small Feature (if any): [BGL_for_lcc](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/BGL_for_lcc)

### CODE REVIEW

File: BGL/doc/BGL/dependencies [hunk 1]
```diff
@@ -16,3 +16,5 @@ Property_map
 Nef_3
 Miscellany
 Polygon_mesh_processing
+Combinatorial_map
+Linear_cell_complex
```
Comment (lrineau) [inline BGL/doc/BGL/dependencies]: The last line should terminate with an end-of-line.

Comment (gdamiand) [inline BGL/doc/BGL/dependencies]: Done.

File: BGL/doc/BGL/graph_traits.txt [hunk 2]
```diff
@@ -84,6 +84,66 @@ The traits class `boost::graph_traits< CGAL::Polyhedron_3<T> >` provides the fol
 For convenience, the type `edge_descriptor` is hashable using the functor `CGAL::Handle_hash_function`
 that is the default hash functor of `CGAL::Unique_hash_map`.
 
+\cgalModifBegin
```
Comment (sloriot) [inline BGL/doc/BGL/graph_traits.txt]: do you want to keep that?

File: BGL/examples/BGL_LCC/normals.cmd [hunk 3]
```diff
@@ -0,0 +1 @@
+cube.off
```
Comment (sloriot) [inline BGL/examples/BGL_LCC/normals.cmd]: seems to me that all the cmd files added should end with _lcc other the test script won't use them.

File: BGL/test/BGL/CMakeLists.txt [hunk 4]
```diff
@@ -87,7 +89,9 @@ create_single_source_cgal_program( "test_Has_member_id.cpp" )
 
 create_single_source_cgal_program( "test_cgal_bgl_named_params.cpp" )
 
-
+if(OpenMesh_FOUND)
+  target_link_libraries( test_graph_geometry ${OPENMESH_LIBRARIES})
+endif()
```
Comment (sloriot) [inline BGL/test/BGL/CMakeLists.txt]: it seems this block is already below. Is it really needed?

File: BGL/test/BGL/test_Euler_operations.cpp [hunk 5]
```diff
@@ -381,9 +381,10 @@ int main()
 {
   test_Euler_operations<Polyhedron>();
   test_Euler_operations<SM>();
+  test_Euler_operations<LCC>();
 
 #ifdef CGAL_USE_OPENMESH
-  test_Euler_operations<OMesh>();
+  //  test_Euler_operations<OMesh>();
```
Comment (sloriot) [inline BGL/test/BGL/test_Euler_operations.cpp]: you probably want to uncomment this.

File: BGL/doc/BGL/graph_traits.txt (hunk 2 above)
Comment (gdamiand) [inline BGL/doc/BGL/graph_traits.txt]: Of course no! Removed.

File: BGL/examples/BGL_LCC/normals.cmd (hunk 3 above)
Comment (gdamiand) [inline BGL/examples/BGL_LCC/normals.cmd]: Done.

File: BGL/test/BGL/CMakeLists.txt (hunk 4 above)
Comment (gdamiand) [inline BGL/test/BGL/CMakeLists.txt]: Done (removed).

File: BGL/test/BGL/test_Euler_operations.cpp (hunk 5 above)
Comment (gdamiand) [inline BGL/test/BGL/test_Euler_operations.cpp]: Done.

File: Linear_cell_complex/include/CGAL/boost/graph/cgal_bgl_graph_io.h [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+    reads the graph `g` from data in the OFF format. Ignores comment lines which start with a hash, and lines with whitespace.
+    \sa Overloads of this function for specific models of the concept `FaceGraph`.
+    \pre The data must represent a 2-manifold
+    \attention The graph `g` is not cleared, and the data from the stream are added.
+
+  */ 
+template <typename FaceGraph>
+bool read_off(std::istream& is,
+              FaceGraph& g)
+{
```
Comment (sloriot) [inline Linear_cell_complex/include/CGAL/boost/graph/cgal_bgl_graph_io.h]: Any reason not to use the generic version?

Comment (afabri) [inline Linear_cell_complex/include/CGAL/boost/graph/cgal_bgl_graph_io.h]: The non-generic version may be capable to read binary, CNOFF, etc

Comment (lrineau) [inline Linear_cell_complex/include/CGAL/boost/graph/cgal_bgl_graph_io.h]: This *is* the generic version, a copy-paste of the file introduced in 2281f83664718cd8dfba2911169c5cccf2e2eaa7 by Andreas in PR #2101.

Comment (gdamiand) [inline Linear_cell_complex/include/CGAL/boost/graph/cgal_bgl_graph_io.h]: Yes this is a copy of the version of Andreas. 
This is of course temporary and will be removed when the small feature will be integrated.

**Comment by lrineau** [comment]
This PR conflicts on three files with https://github.com/CGAL/cgal/pull/1917, already merged in `integration`.

I see no easy way to deal with that situation.

Guillaume, I suggest you try to merge the branch in `integration` yourself, and deal with the conflict. That way the branch would be tested tonight.

**Comment by gdamiand** [comment]
Done.

**Comment by lrineau** [comment]
@lrineau, in https://github.com/CGAL/cgal/pull/2174#issuecomment-311374414:
> Guillaume, I suggest you try to merge the branch in `integration` yourself, and deal with the conflict. That way the branch would be tested tonight.


@gdamiand,  https://github.com/CGAL/cgal/pull/2174#issuecomment-311382388
> Done

However, the HEAD of this PR, https://github.com/CGAL/cgal/pull/2174/commits/013f4473957a5a4f0e8bb1430c115f79d26d0de7, is not in integration. You have merged the older version https://github.com/CGAL/cgal/pull/2174/commits/274113803518fff0a235d927b23e68f1bba09a16.

That is sufficient for the first tests, but that will not be sufficient for the integration of the PR in `master`.

**Comment by gdamiand** [comment]
It is because I use github to solve a conflict. Github merge with master; and I don't have this commit in my branch.
How can I update my branch on cnrs repo to be the same than the one on mine repo ?

**Comment by lrineau** [comment]
Those commands should update your local branch:
```shell
git checkout BGL-LCC-gdamiand
git fetch mine
git merge mine/BGL-LCC-gdamiand
```
The merge will be a fast-forward merge.

**Comment by lrineau** [comment]
I have two warnings in my local build of `integration`, that are related to this PR:
```
/home/lrineau/Git/cgal/Linear_cell_complex/include/CGAL/boost/graph/cgal_bgl_graph_io.h: In function ‘bool CGAL::internal::read_off::is_whitespace(const string&)’:
/home/lrineau/Git/cgal/Linear_cell_complex/include/CGAL/boost/graph/cgal_bgl_graph_io.h:115:20: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     for(int i=0; i < s.size(); i++){
                  ~~^~~~~~~~~~
/home/lrineau/Git/cgal/Linear_cell_complex/include/CGAL/boost/graph/cgal_bgl_graph_io.h: In function ‘bool CGAL::read_off(std::istream&, FaceGraph&)’:
/home/lrineau/Git/cgal/Linear_cell_complex/include/CGAL/boost/graph/cgal_bgl_graph_io.h:151:68: warning: typedef ‘face_descriptor’ locally defined but not used [-Wunused-local-typedefs]
   typedef typename boost::graph_traits<FaceGraph>::face_descriptor face_descriptor;
                                                                    ^~~~~~~~~~~~~~~
```

**Comment by gdamiand** [comment]
Thanks @lrineau . It should be ok now.

About the warnings, they are related to https://github.com/afabri/cgal/tree/BGL-read_write_off-GF ;)

**Comment by lrineau** [comment]
There are compilation errors about `is_valid()` not in OpenMesh model of graphs.
https://cgal.geometryfactory.com/CGAL/testsuite/results-4.11-Ic-52.shtml#BGL

Maybe there is a `is_valid()` global function, overloaded for models having such a feature. @sloriot, @afabri, do you know?

**Comment by lrineau** [comment]
There is also that test failure.

https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.11-Ic-52/Generalized_map/TestReport_lrineau_Fedora-32-Release.gz

Could it be related to this PR? Anyway, that is a test of yours, @gdamiand.

**Comment by sloriot** [comment]
I know it is rather messy right now. There is also `is_valid_polygon_mesh()` but I don't know when we should use them. I remember that I got some compilation error with `is_valid()` or `is _valid_polygon_mesh()` because one was using `operator<<()` on its descriptors which  is clearly not working for `Polyhedron_3` for example. IMO, only  `is_valid_polygon_mesh()` should exist and it should be cleaned up and documented.

**Comment by lrineau** [comment]
I found a third error:
```
------------------------------------------------------------------
- ProgramOutput.edge_collapse_linear_cell_complex.ArchLinux
------------------------------------------------------------------
CGAL::Random()::get_seed() = 1498631508
Usage: simplification_Linear_cell_complex inofffile [outofffile]
```
https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.11-Ic-52/Surface_mesh_simplification_Examples/TestReport_lrineau_ArchLinux.gz

That example program should have a data file and a `.cmd` file, so that it can be launched successfully by the testsuite.

**Comment by lrineau** [comment]
I found a fourth error:
```
In file included from /mnt/testsuite/include/CGAL/boost/graph/graph_traits_Linear_cell_complex_for_combinatorial_map.h:35,
                 from /home/cgal_tester/build/src/cmake/platforms/CentOS6/test/Surface_mesh_segmentation_Examples/segmentation_from_sdf_values_LCC_example.cpp:3:
/mnt/testsuite/include/CGAL/boost/graph/cgal_bgl_graph_io.h: In function ‘bool CGAL::read_off(const char*, FaceGraph&) [with FaceGraph = LCC]’:
/home/cgal_tester/build/src/cmake/platforms/CentOS6/test/Surface_mesh_segmentation_Examples/segmentation_from_sdf_values_LCC_example.cpp:22:   instantiated from here
/mnt/testsuite/include/CGAL/boost/graph/cgal_bgl_graph_io.h:111: error: ‘CGAL::internal::read_off’ is not a function,
/mnt/testsuite/include/CGAL/boost/graph/cgal_bgl_graph_io.h:210: error:   conflict with ‘template<class FaceGraph> bool CGAL::read_off(const char*, FaceGraph&)’
/mnt/testsuite/include/CGAL/boost/graph/cgal_bgl_graph_io.h:214: error:   in call to ‘read_off’
```
in the CentOS6 column.

https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.11-Ic-52/Surface_mesh_segmentation_Examples/TestReport_gimeno_CentOS6.gz

**Comment by lrineau** [comment]
A fifth error, in the documentation this time: because of 13ed160, a file has been renamed, and Doxygen can no longer find it:
```
/home/cgal-testsuite/cgal_doc_build/CGAL-4.11-Ic-52/doc/Convex_hull_3/examples.txt:6: Warning: included file Convex_hull_3/dynamic_hull_3.cpp is not found. Check your EXAMPLE_PATH
citelist:10: Warning: Unexpected new line character
/home/cgal-testsuite/cgal_doc_build/CGAL-4.11-Ic-52/doc/Convex_hull_3/Convex_hull_3.txt:137: Warning: included file Convex_hull_3/dynamic_hull_3.cpp is not found. Check your EXAMPLE_PATH
```
https://cgal.geometryfactory.com/CGAL/Manual_doxygen_test/CGAL-4.11-Ic-52/Convex_hull_3.log

**Comment by gdamiand** [comment]
Many errors in [this platform](https://cgal.geometryfactory.com/CGAL/testsuite/results-4.11-Ic-52.shtml#platform29) about partial specialization of ‘struct boost::is_enum seems related to [this](https://svn.boost.org/trac10/ticket/12534) bug in boost 1.62.

@lrineau How can we deal with this problem ?

**Comment by gdamiand** [comment]
Error about missing dynamic_hull_3.cpp is corrected by https://github.com/CGAL/cgal/pull/2174/commits/8059f828827dfc1bc5bcb7ba5affa126ff5d843a

**Comment by lrineau** [comment]
Debian-testing is stuck with Boost 1.62 [for a while][1]. I do not know why.

  [1]: https://tracker.debian.org/pkg/boost-defaults

We can:
  - ignore the errors for now (*that is what I do*), and wait for Debian to update their default version of Boost to a later Boost version,
  - or try to modify the order of includes, as suggested by [the bug comments][2],
  - or try the [Boost-1.63 packages][3], but I have no idea who they mix with the other packages of the distribution.

  [2]: https://svn.boost.org/trac10/ticket/12534
  [3]: https://tracker.debian.org/pkg/boost1.63

**Comment by gdamiand** [comment]
error: ‘CGAL::internal::read_off’ is solved in https://github.com/CGAL/cgal/pull/2174/commits/d2f80b3da931e0bd6245f21b9f1e115604ced13c (and must be done also in #2101)

**Comment by gdamiand** [comment]
third error is corrected by https://github.com/CGAL/cgal/pull/2174/commits/25d6abcc1a0e06bfd44adab86f0ffeb27162fa6a

**Comment by gdamiand** [comment]
No idea how to solve the first error

error: use of class template 'optional' requires template arguments ?

Is it really related to this PR ?

**Comment by gdamiand** [comment]
The last error in https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.11-Ic-52/Generalized_map/TestReport_lrineau_Fedora-32-Release.gz about copy of generalized map is not related to this PR (nothing was changed in GMap).

I observed this error long time ago; having no clue how to solve it; and the error disappear without any modification from my side. Let us ignore it from now and see if the same magic will hapen also this time ?

**Comment by lrineau** [comment]
The is still a build error in my local build:
```
In file included from /usr/include/c++/6.3.1/cassert:44:0,
                 from /usr/include/OpenMesh/Core/Mesh/CirculatorsT.hh:62,
                 from /usr/include/OpenMesh/Core/Mesh/PolyConnectivity.hh:54,
                 from /usr/include/OpenMesh/Core/Mesh/PolyMesh_ArrayKernelT.hh:65,
                 from /home/lrineau/Git/cgal/BGL/test/BGL/test_Prefix.h:52,
                 from /home/lrineau/Git/cgal/BGL/test/BGL/test_graph_traits.cpp:3:
/home/lrineau/Git/cgal/BGL/test/BGL/test_Prefix.h: In instantiation of ‘std::vector<T> t_data() [with T = OpenMesh::PolyMesh_ArrayKernelT<>]’:
/home/lrineau/Git/cgal/BGL/test/BGL/test_Prefix.h:167:24:   required from here
/home/lrineau/Git/cgal/BGL/test/BGL/test_Prefix.h:151:14: error: ‘class OpenMesh::PolyMesh_ArrayKernelT<>’ has no member named ‘is_valid’
     assert(s.is_valid());
            ~~^
```

**Comment by gdamiand** [comment]
Corrected https://github.com/CGAL/cgal/pull/2174/commits/af704704e7c10c4c37c3212461c67303ff25628b

**Comment by lrineau** [comment]
@gdamiand : That is a lot better in CGAL-4.11-Ic-53, but this PR has redden the [`Polygon_mesh_processing/test`] line, with that error everywhere:
```
- ProgramOutput.surface_intersection_sm_poly.Debian-Stable
------------------------------------------------------------------
CGAL::Random()::get_seed() = 1498683189
data/XXXXXXX.off is not a valid off file.
```

[`Polygon_mesh_processing/test`]: https://cgal.geometryfactory.com/CGAL/testsuite/results-4.11-Ic-53.shtml#Polygon_mesh_processing

(Cc @afabri )

**Comment by gdamiand** [comment]
I think I am not guilty for this one: it seems to have the same problem in master (file data/XXXXXXX.off does not exist and there is no surface_intersection_sm_poly.cmd).

Strange...

**Comment by gdamiand** [comment]
@lrineau Error about GMap copy is gone (compare [4.11-Ic-52](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.11-Ic-52/Generalized_map/TestReport_lrineau_Fedora-32-Release.gz) and [4.11-Ic-53](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.11-Ic-53/Generalized_map/TestReport_lrineau_Fedora-32-Release.gz)) without any reason.

I think this is a bug in the compiler...

**Comment by gdamiand** [comment]
For the error about partial specialization of ‘struct boost::is_enum, I tried to move the <code>#include <boost/container/flat_map.hpp></code> in first. Let us see if it solves the problem.

(If yes this must also be done in PR #2101 ).

**Comment by sloriot** [comment]
About `surface_intersection_sm_poly.cpp` it is a mistake. I forgot to put real filenames. `data/blobby.off`   and `data/eight.off` are working. I can make the fix if you want.

**Comment by gdamiand** [comment]
Thanks @sloriot .
I can do this in this branch and it will solve the problem in master when it will be integrated.
@lrineau Is it ok for you ?

**Comment by lrineau** [comment]
Le Thursday, June 29, 2017 4:10:07 PM CEST Sebastien Loriot a écrit :
> About `surface_intersection_sm_poly.cpp` it is a mistake. I forgot to put
> real filenames. `data/blobby.off`   and `data/eight.off` are working. I
> can make the fix if you want.

Le Thursday, June 29, 2017 4:11:13 PM CEST Guillaume Damiand a écrit :
> Thanks @sloriot .
> I can do this in this branch and it will solve the problem in master when
> it will be integrated. @lrineau Is it ok for you ?

I agree.

-- 
Laurent Rineau, PhD
R&D Engineer at GeometryFactory           http://www.geometryfactory.com/
Release Manager of the CGAL Project       http://www.cgal.org/

**Comment by lrineau** [comment]
The bug of Boost about the partial specialization of `boost::is_enum<std::pair<..>>` is really annoying.

I propose a fix, in `<CGAL/config.h>`:
- as soon as `<boost/config.hpp>` is included, tested if the version if 1.62, if it is: include a Boost Container header, like `<boost/container/flat_map.hpp>`, directly in `<CGAL/config.h>`. That should be early enough.

**Comment by lrineau** [comment]
I have verified in Boost git repo that the bug was introduced in Boost-1.62 and fixed in 1.63. I will implement the patch I propose.

**Comment by lrineau** [comment]
Done in https://github.com/CGAL/cgal/pull/2174/commits/b0336e655859783d02855a265ef879263c11e7ef.

**Comment by gdamiand** [comment]
Thanks Laurent !

**Comment by lrineau** [comment]
This PR is now merged.

The `changes.html` has not been changed. Is that wanted?

**Comment by gdamiand** [comment]
Sorry; an oversight. Updated in #2255

**Comment by lrineau** [comment]
https://github.com/CGAL/cgal/pull/2101 is now merged in `master`. The I/O should now be updated.

**Comment by gdamiand** [comment]
@afabri Replacing my copy of your file by the version in master, I have errors about undefined function degree (for write_off function). Do you also have these errors ? Where this function is defined ?

**Comment by lrineau** [comment]
`degree` is in the concept of [FaceGraph][*]. It is to be defined by all model of that concept. However, it was not tested, so far, and that may explain why it was forgot from the LCC BGL adapters.

  [*]: https://cgal.geometryfactory.com/CGAL/doc/master/BGL/classFaceGraph.html

**Comment by gdamiand** [comment]
In https://github.com/CGAL/cgal/pull/2101, I suggested to add the generic version of degree; because I don't think this is possible to really optimize it for the different models of FaceGraph...

**Pull request merged**